Wrap this tool in a container image

vinman1 opened the pull request:
Using the new `docker/Dockerfile`, create a container image where `markdownlint-cli2` is installed globally and is also set as the default entrypoint. This makes usage of this image for linting MD files identical to using `markdownlint-cli2` directly:

```bash
$ docker run --rm -v $PWD:/workdir -w /workdir DavidAnson/markdownlint-cli2:latest "**/*.md" "#node_modules"
```

2 new scripts were added to the `package.json`. Both of these new scripts assume you have [Docker](https://www.docker.com/) installed:

1. `build-container`: which uses `docker` to build the container and tag it with `latest` as well as with a version defined by `$SEMVAR`

    Call it like so (or find a way to automate setting the `$SEMVAR` before calling it)

    ```bash
    $ SEMVAR=0.2.0 npm run build-container
    ```

    The above command will produce an image with 2 tags:
    - `DavidAnson/markdownlint-cli2:0.2.0`
    - `DavidAnson/markdownlint-cli2:latest`

2. `lint-dockerfile`: which uses a very popular `Dockerfile` linter, [`hadolint`](https://github.com/hadolint/hadolint), to lint `docker/Dockerfile`.

    You can call it on its own like so:
    ```bash
    $ npm run lint-dockerfile
    ```
   but I also added it to the existing `ci` script:
    ```bash
    $ npm run ci
    ```

What's remaining is the publication of the built image (both tags) to a public registry. Obviously I would suggest https://hub.docker.com for the most exposure. DockerHub allows any free account unlimited _public_ repositories (images) and only 1 _private_ one by default. In this case, we would want the created repository to be _public_.

For now, the above `build-container` script assumes your username on DockerHub is `DavidAnson` and therefore prefixes the image with it (a standard DockerHub convention). If your username is anything other than `DavidAnson` you will need to update the `build-container` script to prefix the tags accordingly.

Once you've settled on a username and updated the `build-container` script accordingly, a new publication script could be added to the `package.json`, e.g.:

```json
  "scripts":
    ...
    "publish-container": "docker push DavidAnson/markdownlint-cli2:$SEMVAR && docker push DavidAnson/markdownlint-cli2:latest"
```

The above script makes the following assumptions:

1. You've previously called the `build-container` script which created/updated the image tags `$SEMVAR` and `latest`.
2. You're already logged in (authenticated) to DockerHub (done by calling `docker login`).
3. You will also call this `publish-container` script by providing the `$SEMVAR` variable, e.g.:

    ```bash
    $ SEMVAR=0.2.0 npm run publish-container
    ```

4. Again, that your username on DockerHub is `DavidAnson` (update the script otherwise)

If you're so inclined, you might want to add some labels to the built container image (see the [Dockerfile LABEL command reference](https://docs.docker.com/engine/reference/builder/#label)).

This PR closes #27.

DavidAnson commented:
This looks very thorough, thank you!! I'm in the middle of a cross-repo change, but will look at this PR next.

FYI, I created a Docker Hub account as "davidanson" because they do not allow capital letters.

vinman1 commented:
I've updated the container-image prefix to `davidanson` and I've added the `publish-container` script to the `package.json`.

Now I'm looking for an easy way to include the `lint-dockerfile` script so that its callable both locally and via your configured GitHub workflow.

vinman1 commented on .github/workflows/ci.yml:
```diff
@@ -28,6 +28,14 @@ jobs:
+        LATEST_RELEASE=$(curl -L -s -H "Accept: application/json" "https://github.com/hadolint/hadolint/releases/latest" | jq -r '.tag_name')
+        curl -sL -o hadolint "https://github.com/hadolint/hadolint/releases/download/${LATEST_RELEASE}/hadolint-Linux-x86_64"
+        sudo chmod +x hadolint
+        ./hadolint docker/Dockerfile 
```
depending on how strict you are with github workflow definition, there might be a simpler way to do this. If you agree with having a separate job to perform this step, we could use the available container image instead of downloading and installing `hadolint` (I'm less familiar with github workflow syntax...):

```yaml
jobs:
  lint-dockerfile:
    runs-on: ubuntu-latest
    container: hadolint/hadolint:latest-alpine
    steps:
    - uses: actions/checkout@v2
    - name: hadolint
    - run: hadolint docker/Dockerfile

```

DavidAnson replied:
I like the dedicated job idea. Looks much cleaner. :) I’ve used that in a couple of other places.

DavidAnson commented on package.json:
```diff
@@ -21,10 +21,13 @@
   "scripts": {
     "ci": "npm-run-all --continue-on-error --parallel test-cover lint",
     "lint": "eslint --max-warnings 0 .",
+    "lint-dockerfile": "docker run --rm -i hadolint/hadolint:latest-alpine < docker/Dockerfile",
```
I like this. But I'm not sure about also putting it in the "ci" script because then everyone needs to have Docker installed. Currently, it's just Node that's required. I may prefer to rely on the GitHub action for this. Thoughts?

DavidAnson replied:
Oh, you didn't add it there. Maybe we agree.

vinman1 replied:
I had originally put it the ci script so that it got linted at the same time as the rest. However, then as you said, it introduced the requirement to have Docker installed. So I've kept it separate.

DavidAnson commented on package.json:
```diff
@@ -21,10 +21,13 @@
     "test-cover": "c8 --check-coverage --branches 100 --functions 100 --lines 100 --statements 100 npm test",
-    "test-watch": "git ls-files | entr npm run test"
+    "test-watch": "git ls-files | entr npm run test",
+    "build-container": "docker build -t davidanson/markdownlint-cli2:$SEMVAR -f docker/Dockerfile . && docker tag davidanson/markdownlint-cli2:$SEMVAR davidanson/markdownlint-cli2:latest",
```
Please sort alphabetically.

DavidAnson commented on package.json:
```diff
@@ -21,10 +21,13 @@
-    "test-watch": "git ls-files | entr npm run test"
+    "test-watch": "git ls-files | entr npm run test",
+    "build-container": "docker build -t davidanson/markdownlint-cli2:$SEMVAR -f docker/Dockerfile . && docker tag davidanson/markdownlint-cli2:$SEMVAR davidanson/markdownlint-cli2:latest",
+    "publish-container": "docker push davidanson/markdownlint-cli2:$SEMVAR && docker push davidanson/markdownlint-cli2:latest"
```
I'm not sure I love SEMVAR. What about a Node one-liner to pull the version out of package.json instead? Something line `node -e "require('package.json').version"` (from memory on iPad)?

vinman1 replied:
I like it. I'll try that.

vinman1 replied:
got it working, it looks something like this:

```json
  "scripts": {
    "build-container": "VERSION=$(node -e \"process.stdout.write(require('./package.json').version)\") && docker build -t davidanson/markdownlint-cli2:$VERSION -f docker/Dockerfile . && docker tag davidanson/markdownlint-cli2:$VERSION davidanson/markdownlint-cli2:latest",
    "publish-container": "VERSION=$(node -e \"process.stdout.write(require('./package.json').version)\") && docker push davidanson/markdownlint-cli2:$VERSION && docker push davidanson/markdownlint-cli2:latest",
  },
```

DavidAnson reviewed (changes requested):
I made a first pass review after reading the Docker docs. Sorry for a lot of beginner questions!

DavidAnson commented on docker/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+FROM node:lts-alpine AS build
```
(The "docker" directory)

vinman1 replied:
yup, completely slipped my mind. Good catch!

vinman1 replied:
Actually, I just realized the project doesn't have an `.npmignore`. Instead the files to include in the package are explicitly specified in the `package.json`:
```json
  "files": [
    "append-to-array.js",
    "markdownlint-cli2.js",
    "markdownlint-cli2-fix.js",
    "merge-options.js",
    "resolve-and-require.js"
  ],
```

the `npm pack` command additionally includes `README.md`, `LICENSE`, and `package.json` automatically. 

Because of this, I don't think the project needs a `.npmignore` file. Have I understood correctly?

vinman1 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+FROM node:lts-alpine AS build
+WORKDIR /pack
+COPY --chown=1000 . .
```
I had done this as a precaution. Here's my reasoning:

There are 2 users included in the `node:lts-alpine` container image: 
1. The default `root` user, which has full admin rights/permissions
2. A `node`  user (with UID=1000 and GID=1000) who has more limited permissions, but can execute node.

By default, the COPY command would add everything with `root` ownership, which means the `node` user might not be able to access everything. 

Since this step is done only to _pack_ the library, it might not be required... I'll take a closer look.

vinman1 replied:
Ok. so it doesn't seem to have an effect, so I'll remove it.

vinman1 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+COPY --chown=1000 . .
+RUN npm pack
+
+FROM node:lts-alpine AS prod
```
You're right, this `prod` as clause is not referenced anywhere.

vinman1 commented on .github/workflows/ci.yml:
```diff
@@ -8,6 +8,14 @@ on:
+    container: hadolint/hadolint:latest-alpine
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use hadolint to lint Dockerfile
```
ok. will remove

vinman1 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+RUN npm install --global --no-package-lock --production markdownlint-cli2-*.tgz && \
+    rm /markdownlint-cli2-*.tgz
+
+USER node
```
As I mentioned in the above comment, there are 2 users included in the `node:lts-alpine` container image:

The default `root` user, which has full admin rights/permissions
A `node` user who has more limited permissions, but can execute node.

By default, without this `USER node` statement, the container would run as `root`, which is a security risk. It is a common best practice that containers execute under a non-sudo user, which `node` is. So this statement tells the container to execute all following commands (including the default entrypoint) as the `node` user.

vinman1 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+RUN npm pack
+
+FROM node:lts-alpine AS prod
+# hadolint ignore=DL3010
```
If you're refering to the hadolint rule `DL3010` which I explicitly ignored here, the reason is because the linter mis-interpreted my intention.

As a more general rule, `COPY` is preferred over `ADD`. However, the rule `DL3010` makes an assumption that if I'm copying a `*.tgz` file into a container image, I want to extract it too. The `ADD` command offers a special feature when adding `*.tgz` files where it can automatically extract its contents for you. The `COPY` does not offer the same feature, it just copies. That is why `hadolint` suggests that I use `ADD` instead of `COPY`. 

However in our specific usage here, we don't want to extract the contents of the *.tgz ourselves; we just want to copy the file into the container. Our intention is to let the `npm install` command we execute next take care of the extraction/installation for us. That is why I've explicitly ignored this rule on this line.

vinman1 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+# hadolint ignore=DL3010
+COPY --from=build /pack/markdownlint-cli2-*.tgz /
+RUN npm install --global --no-package-lock --production markdownlint-cli2-*.tgz && \
+    rm /markdownlint-cli2-*.tgz
```
Honestly, I hadn't considered the impact on the image size. My only thought was to cleanup after the operation completed.

Now that I think about it, since the `rm` operation is performed as part of the same `RUN` command as the `npm install`, it will not add another layer. Each _docker_ command (COPY, ADD, RUN, USER, etc...)  in a Dockerfile adds a new layer. But multiple commands executed as part of a single `RUN` command are all included in the same layer.

Which means that in theory, the final image size should be smaller, because this layer adds less bytes to the filesystem. In practice, however, because the `markdownlint-cli2-0.2.0.tgz` file is only about 12KB, its inclusion/exclusion makes no perceivable difference to the image size.

So I think I should leave this as is, for cleanliness and to follow best practices. Agreed?

vinman1 commented on docker/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+
+USER node
+
+ENTRYPOINT ["/usr/local/bin/markdownlint-cli2"]
```
So I see that, yes, both `markdownlint-cli2` and `markdownlint-cli2-fix` are available in the resulting container. However, by definition, a container can only have a single `ENTRYPOINT`.

Since this line sets the default `ENTRYPOINT` as `markdownlint-cli2`, a user that would want to call the `markdownlint-cli2-fix` instead would have to execute the container by explicitly specifying it as the entrypoint; like this:

```bash
$ docker run --rm -v $PWD:/app -w /app --entrypoint="markdownlint-cli2-fix" davidanson/markdownlint-cli2 "**/*.md" "#node_modules"
```

As an alternative, we _could_ have the container execute `markdownlint-cli2-fix` instead of `markdownlint-cli2`  if the user passed a  `--fix` command line arg; something like:

```bash
$ docker run --rm -v $PWD:/app -w /app davidanson/markdownlint-cli2:latest --fix "**/*.md" "#node_modules"
```
However, the `--fix` argument would then be passed along to the tool:
```bash
markdownlint-cli2 v0.2.0 (markdownlint v0.24.0)
Finding: --fix **/*.md !node_modules
...
```
Which from my understanding will look for a file/directory called `--fix` to lint.

IMHO, stick with explicitly specifying the `ENTRYPOINT` for now. If you eventually want the container to offer both _as a single tool_, then `markdownlint-cli2` itself should also offer it _as a single tool_. I mean, instead of providing a separate `markdownlint-cli2-fix` executable, incorporate it as an argument that you pass to `mardownlint-cli2` itself. Many other linters do it this way (e.g. prettier.js). This could be a future issue/PR if you deem it appropriate.

What we should do, is update the README.md to include docker usage scenarios, including this one to call `markdownlint-cli2-fix`.
